Src/Newtonsoft.Json/JsonTextReader.Async.cs

```diff
@@ -0,0 +1,2327 @@
+                            default:
+                                _charPos = charPos;
+                                throw JsonReaderException.Create(this, "Bad JSON escape sequence: {0}.".FormatWith(CultureInfo.InvariantCulture, @"\" + currentChar));
+                        }
+
+                        EnsureBufferNotEmpty();
+                        WriteCharToBuffer(writeChar, lastWritePosition, escapeStartPos);
+
+                        lastWritePosition = charPos;
+                        break;
+                    case StringUtils.CarriageReturn:
+                        _charPos = charPos - 1;
+                        await ProcessCarriageReturnAsync(true, cancellationToken).ConfigureAwait(false);
+                        charPos = _charPos;
+                        break;
+                    case StringUtils.LineFeed:
+                        _charPos = charPos - 1;
+                        ProcessLineFeed();
+                        charPos = _charPos;
+                        break;
+                    case '"':
+                    case '\'':
+                        if (_chars[charPos - 1] == quote)
+                        {
+                            charPos--;
+
+                            if (initialPosition == lastWritePosition)
+                            {
+                                _stringReference = new StringReference(_chars, initialPosition, charPos - initialPosition);
+                            }
+                            else
+                            {
+                                EnsureBufferNotEmpty();
+
+                                if (charPos > lastWritePosition)
+                                {
+                                    _stringBuffer.Append(_arrayPool, _chars, lastWritePosition, charPos - lastWritePosition);
+                                }
+
+                                _stringReference = new StringReference(_stringBuffer.InternalBuffer, 0, _stringBuffer.Position);
+                            }
+
+                            charPos++;
+                            _charPos = charPos;
+                            return;
+                        }
+
+                        break;
+                }
+            }
+        }
+
+        private async Task ProcessCarriageReturnAsync(bool append, CancellationToken cancellationToken)
+        {
+            _charPos++;
+
+            if (await EnsureCharsAsync(1, append, cancellationToken).ConfigureAwait(false) && _chars[_charPos] == StringUtils.LineFeed)
+            {
+                _charPos++;
+            }
+
+            OnNewLine(_charPos);
+        }
+
+        private async Task<char> ParseUnicodeAsync(CancellationToken cancellationToken)
+        {
+            char writeChar;
+            if (await EnsureCharsAsync(4, true, cancellationToken).ConfigureAwait(false))
+            {
+                char hexChar = Convert.ToChar(ConvertUtils.HexTextToInt(_chars, _charPos, _charPos + 4));
```

Comment:
Can move this plus the next 3 lines into a method that can be used by both `ParseUnicode`s